src/pvcam.cpp

```diff
@@ -0,0 +1,1200 @@
+    query_pixel_type_capabilities(meta);
+}
+
+void
+PVCamCamera::get_shape(ImageShape* shape) const
+{
+    CHECK(shape);
+
+    const auto width =
+      last_known_settings_.shape.x / last_known_settings_.binning;
+    const auto height =
+      last_known_settings_.shape.y / last_known_settings_.binning;
+
+    *shape = {
+                .dims = {
+                  .channels = 1,
+                  .width = (uint32_t)width,
+                  .height = (uint32_t)height,
+                  .planes = 1,
+                },
+                .strides = {
+                  .channels = 1,
+                  .width = 1,
+                  .height = width,
+                  .planes = width * height,
+                },
+                .type = last_known_settings_.pixel_type,
+            };
+}
+
+void
+PVCamCamera::start()
+{
+    using namespace std::chrono_literals;
+
+    CHECK(!started_);
+
+    // allocate internal frame buffers, counters, and cursors
+    uint64_t frame_buffer_size;
+    PVCAM(pl_get_param(
+            hcam_, PARAM_FRAME_BUFFER_SIZE, ATTR_DEFAULT, &frame_buffer_size),
+          *pvcam_api_mutex_);
+
+    pl_buffer_size_ = frame_buffer_size;
+    pl_buffer_ = new uint8_t[pl_buffer_size_];
+
+    nframes_buf_ = (uint32_t)(frame_buffer_size / bytes_per_frame_);
+    frame_count_ = 0;
+
+    // start acquisition
+    PVCAM(
+      pl_exp_start_cont(hcam_, (void*)pl_buffer_, (uint32_t)pl_buffer_size_),
+      *pvcam_api_mutex_);
+
+    int16_t status;
+    uint32_t byte_cnt, buffer_cnt;
+    PVCAM(pl_exp_check_cont_status(hcam_, &status, &byte_cnt, &buffer_cnt),
+          *pvcam_api_mutex_);
+
+    started_ = true;
+}
+
+void
+PVCamCamera::stop()
+{
+    if (!started_) {
+        return;
+    }
+
+    LOG("Stopping");
+    PVCAM(pl_exp_stop_cont(hcam_, CCS_HALT), *pvcam_api_mutex_);
+    started_ = false;
+
+    delete[] pl_buffer_;
+    pl_buffer_ = nullptr;
+    pl_buffer_size_ = 0;
+}
+
+void
+PVCamCamera::execute_trigger() const
+{
+    // In current implementation the input flags should be set to 0. On output,
+    // the flags will contain one of the values defined in PL_SW_TRIG_STATUSES
+    // enumeration.
+    uint32_t flags = 0;
+    PVCAM(pl_exp_trigger(hcam_, &flags, 0), *pvcam_api_mutex_);
+
+    if (flags != PL_SW_TRIG_STATUS_TRIGGERED) {
+        LOGE("The camera was unable to accept the trigger due to an ongoing "
+             "exposure.");
+    }
+}
+
+void
+PVCamCamera::get_frame(void* im, size_t* nbytes, struct ImageInfo* info)
+{
+    CHECK(im);
+    CHECK(nbytes);
+    CHECK(info);
+
+    int16_t status;
+    uint32_t byte_cnt, buffer_cnt;
+    PVCAM(pl_exp_check_cont_status(hcam_, &status, &byte_cnt, &buffer_cnt),
+          *pvcam_api_mutex_);
+
+    std::unique_lock<std::mutex> frame_lock(*frame_mutex_);
+
+    // callback will increment frames_available
+    callback_context_.cv.wait_for(
+      frame_lock,
+      std::chrono::microseconds(exposure_time_ * exposure_time_to_us_),
```

Comment:
I guess the reason we wait for up to the *exact frame time* here is because Acquire is constantly polling for frames at the equivalent frame rate, so this roughly prevents Acquire's source/camera thread from backing up - is that right?

Reply:
That's right, or that was the idea anyway.

Reply:
I guess the timed wait also guards against the callback throwing, so I think I'm OK with this.

However, I feel like this is an important design decision, so it could be worth reviewing this in person and/or getting feedback from @nclack  here.